schema registry maven plugin fails to registers schemas for subjects with slashes "/"

### Schema registry maven plugin fails to register schemas having slashes "/" in subject name even after replacing "/" with "_x2F".  


- See attached zip file with pom.xml (s), protos being used and errors being reported while registering schemas. [schema_registry_issue.zip](https://github.com/confluentinc/schema-registry/files/6819447/schema_registry_issue.zip)

- Issue is consistently reproducible with maven plugin versions 6.1.2 and 6.2.0 

### Identified root cause

It seems that UploadSchemaRegistryMojo.java doesn't use encoded subject names for finding subject internally from subjects, schematypes and references HashMaps.  

Ideally one should use

-  encoded subject names (i.e. subject name which replace "/" with "_x2F") for all internal map lookups i.e. to find subject, reference, schemaTypes (created from provided pom.xml) as these are xml tags and they cannot contain "/"

- and  decoded subject names and references (i.e. with "/") while registering schemas to avoid any exceptions.


### Error Details
e.g. **following pom.xml results into errors.  If you will look carefully subject name tags are already using proposed fix of replacing "/" with "_x2F".**

Command used to register schemas = **mvn schema-registry:register -Dschema-registry-url=http://localhost:8081**

**Pom.xml**

`<project xmlns="http://maven.apache.org/POM/4.0.0"
       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.schema.registry.test</groupId>
  <artifactId>kafka-schemas</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <pluginRepositories>
      <pluginRepository>
          <id>confluent</id>
          <url>https://packages.confluent.io/maven/</url>
      </pluginRepository>
  </pluginRepositories>
  <build>
      <plugins>
          <plugin>
              <groupId>io.confluent</groupId>
              <artifactId>kafka-schema-registry-maven-plugin</artifactId>
              <version>6.2.0</version>
              <configuration>
                  <schemaRegistryUrls>
                      <param>${schema-registry-url}</param>
                  </schemaRegistryUrls>
                  <userInfoConfig>${schema-registry-user-info}</userInfoConfig>
                  <references>
                      <First.proto>
                        <reference>
                            <subject>Second.proto</subject>
                            <name>Second.proto</name>
                        </reference>
                      </First.proto>
    <Second.proto>
      <reference>
          <subject>google/protobuf/timestamp.proto</subject>
          <name>google/protobuf/timestamp.proto</name>
      </reference>
      <reference>
          <subject>google/protobuf/wrappers.proto</subject>
          <name>google/protobuf/wrappers.proto</name>
      </reference>
    </Second.proto>
</references>
<subjects>
    <First.proto>schemas/First.proto</First.proto>
    <google_x2Fprotobuf_x2Fwrappers.proto>schemas/google/protobuf/wrappers.proto</google_x2Fprotobuf_x2Fwrappers.proto>
    <google_x2Fprotobuf_x2Ftimestamp.proto>schemas/google/protobuf/timestamp.proto</google_x2Fprotobuf_x2Ftimestamp.proto>
    <Second.proto>schemas/Second.proto</Second.proto>
</subjects>
                    <schemaTypes>
                        <First.proto>PROTOBUF</First.proto>
                        <google_x2Fprotobuf_x2Fwrappers.proto>PROTOBUF</google_x2Fprotobuf_x2Fwrappers.proto>
                        <google_x2Fprotobuf_x2Ftimestamp.proto>PROTOBUF</google_x2Fprotobuf_x2Ftimestamp.proto>
                        <Second.proto>PROTOBUF</Second.proto>
                    </schemaTypes>
                </configuration>
            </plugin>
        </plugins>
    </build>
</project>`


**Error thrown by schema registry maven plugin**

> [INFO] Scanning for projects...
> [INFO]
> [INFO] ---------------< com.schema.registry.test:kafka-schemas >---------------
> [INFO] Building kafka-schemas 0.0.1-SNAPSHOT
> [INFO] --------------------------------[ jar ]---------------------------------
> [INFO]
> [INFO] --- kafka-schema-registry-maven-plugin:6.2.0:register (default-cli) @ kafka-schemas ---
> [WARNING] Version not specified for ref with name 'google/protobuf/timestamp.proto' and subject 'google/protobuf/timestamp.proto', using latest version
> [WARNING] Version not specified for ref with name 'google/protobuf/wrappers.proto' and subject 'google/protobuf/wrappers.proto', using latest version
> [ERROR] Could not parse Protobuf schema
> java.lang.RuntimeException: io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException: Subject 'google/protobuf/timestamp.proto' not found.; error code: 40401
>     at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getByVersion (CachedSchemaRegistryClient.java:340)
>     at io.confluent.kafka.schemaregistry.AbstractSchemaProvider.resolveReferences (AbstractSchemaProvider.java:60)
>     at io.confluent.kafka.schemaregistry.AbstractSchemaProvider.resolveReferences (AbstractSchemaProvider.java:47)
>     at io.confluent.kafka.schemaregistry.protobuf.ProtobufSchemaProvider.parseSchema (ProtobufSchemaProvider.java:45)
>     at io.confluent.kafka.schemaregistry.SchemaProvider.parseSchema (SchemaProvider.java:63)
>     at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.parseSchema (CachedSchemaRegistryClient.java:204)
>     at io.confluent.kafka.schemaregistry.maven.UploadSchemaRegistryMojo.processSubject (UploadSchemaRegistryMojo.java:105)
>     at io.confluent.kafka.schemaregistry.maven.UploadSchemaRegistryMojo.getReferences (UploadSchemaRegistryMojo.java:153)
>     at io.confluent.kafka.schemaregistry.maven.UploadSchemaRegistryMojo.processSubject (UploadSchemaRegistryMojo.java:95)
>     at io.confluent.kafka.schemaregistry.maven.UploadSchemaRegistryMojo.execute (UploadSchemaRegistryMojo.java:79)
>     at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
>     at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
>     at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
>     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
>     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
>     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
>     at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
>     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
>     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
>     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
>     at java.lang.reflect.Method.invoke (Method.java:566)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
> Caused by: io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException: Subject 'google/protobuf/timestamp.proto' not found.; error code: 40401
>     at io.confluent.kafka.schemaregistry.client.rest.RestService.sendHttpRequest (RestService.java:295)
>     at io.confluent.kafka.schemaregistry.client.rest.RestService.httpRequest (RestService.java:365)
>     at io.confluent.kafka.schemaregistry.client.rest.RestService.getVersion (RestService.java:758)
>     at io.confluent.kafka.schemaregistry.client.rest.RestService.getVersion (RestService.java:742)
>     at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getByVersion (CachedSchemaRegistryClient.java:338)
>     at io.confluent.kafka.schemaregistry.AbstractSchemaProvider.resolveReferences (AbstractSchemaProvider.java:60)
>     at io.confluent.kafka.schemaregistry.AbstractSchemaProvider.resolveReferences (AbstractSchemaProvider.java:47)
>     at io.confluent.kafka.schemaregistry.protobuf.ProtobufSchemaProvider.parseSchema (ProtobufSchemaProvider.java:45)
>     at io.confluent.kafka.schemaregistry.SchemaProvider.parseSchema (SchemaProvider.java:63)
>     at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.parseSchema (CachedSchemaRegistryClient.java:204)
>     at io.confluent.kafka.schemaregistry.maven.UploadSchemaRegistryMojo.processSubject (UploadSchemaRegistryMojo.java:105)
>     at io.confluent.kafka.schemaregistry.maven.UploadSchemaRegistryMojo.getReferences (UploadSchemaRegistryMojo.java:153)
>     at io.confluent.kafka.schemaregistry.maven.UploadSchemaRegistryMojo.processSubject (UploadSchemaRegistryMojo.java:95)
>     at io.confluent.kafka.schemaregistry.maven.UploadSchemaRegistryMojo.execute (UploadSchemaRegistryMojo.java:79)
>     at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
>     at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
>     at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
>     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
>     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
>     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
>     at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
>     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
>     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
>     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
>     at java.lang.reflect.Method.invoke (Method.java:566)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
> [ERROR] Schema for Second.proto could not be parsed.
> [WARNING] Version not specified for ref with name 'Second.proto' and subject 'Second.proto', using latest version
> [ERROR] Could not parse Protobuf schema
> java.lang.RuntimeException: io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException: Subject 'Second.proto' not found.; error code: 40401
>     at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getByVersion (CachedSchemaRegistryClient.java:340)
>     at io.confluent.kafka.schemaregistry.AbstractSchemaProvider.resolveReferences (AbstractSchemaProvider.java:60)
>     at io.confluent.kafka.schemaregistry.AbstractSchemaProvider.resolveReferences (AbstractSchemaProvider.java:47)
>     at io.confluent.kafka.schemaregistry.protobuf.ProtobufSchemaProvider.parseSchema (ProtobufSchemaProvider.java:45)
>     at io.confluent.kafka.schemaregistry.SchemaProvider.parseSchema (SchemaProvider.java:63)
>     at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.parseSchema (CachedSchemaRegistryClient.java:204)
>     at io.confluent.kafka.schemaregistry.maven.UploadSchemaRegistryMojo.processSubject (UploadSchemaRegistryMojo.java:105)
>     at io.confluent.kafka.schemaregistry.maven.UploadSchemaRegistryMojo.execute (UploadSchemaRegistryMojo.java:79)
>     at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
>     at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
>     at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
>     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
>     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
>     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
>     at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
>     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
>     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
>     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
>     at java.lang.reflect.Method.invoke (Method.java:566)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
> Caused by: io.confluent.kafka.schemaregistry.client.rest.exceptions.RestClientException: Subject 'Second.proto' not found.; error code: 40401
>     at io.confluent.kafka.schemaregistry.client.rest.RestService.sendHttpRequest (RestService.java:295)
>     at io.confluent.kafka.schemaregistry.client.rest.RestService.httpRequest (RestService.java:365)
>     at io.confluent.kafka.schemaregistry.client.rest.RestService.getVersion (RestService.java:758)
>     at io.confluent.kafka.schemaregistry.client.rest.RestService.getVersion (RestService.java:742)
>     at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getByVersion (CachedSchemaRegistryClient.java:338)
>     at io.confluent.kafka.schemaregistry.AbstractSchemaProvider.resolveReferences (AbstractSchemaProvider.java:60)
>     at io.confluent.kafka.schemaregistry.AbstractSchemaProvider.resolveReferences (AbstractSchemaProvider.java:47)
>     at io.confluent.kafka.schemaregistry.protobuf.ProtobufSchemaProvider.parseSchema (ProtobufSchemaProvider.java:45)
>     at io.confluent.kafka.schemaregistry.SchemaProvider.parseSchema (SchemaProvider.java:63)
>     at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.parseSchema (CachedSchemaRegistryClient.java:204)
>     at io.confluent.kafka.schemaregistry.maven.UploadSchemaRegistryMojo.processSubject (UploadSchemaRegistryMojo.java:105)
>     at io.confluent.kafka.schemaregistry.maven.UploadSchemaRegistryMojo.execute (UploadSchemaRegistryMojo.java:79)
>     at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
>     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
>     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
>     at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
>     at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
>     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
>     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
>     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
>     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
>     at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
>     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
>     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
>     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
>     at java.lang.reflect.Method.invoke (Method.java:566)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
>     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
> [ERROR] Schema for First.proto could not be parsed.
> [INFO] Registered subject(google/protobuf/timestamp.proto) with id 1 version 1
> [INFO] Registered subject(google/protobuf/wrappers.proto) with id 2 version 1
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  5.394 s
> [INFO] Finished at: 2021-07-14T19:22:08-04:00
> [INFO] ------------------------------------------------------------------------
> [ERROR] Failed to execute goal io.confluent:kafka-schema-registry-maven-plugin:6.2.0:register (default-cli) on project kafka-schemas: Execution default-cli of goal io.confluent:kafka-schema-registry-maven-plugin:6.2.0:register failed: One or more exceptions were encountered. -> [Help 1]
> [ERROR]
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR]
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException

--- a/maven-plugin/src/main/java/io/confluent/kafka/schemaregistry/maven/UploadSchemaRegistryMojo.java
+++ b/maven-plugin/src/main/java/io/confluent/kafka/schemaregistry/maven/UploadSchemaRegistryMojo.java
@@ -19,6 +19,8 @@ package io.confluent.kafka.schemaregistry.maven;
 import com.google.common.base.Preconditions;
 import java.io.UnsupportedEncodingException;
 import java.net.URLDecoder;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
 import org.apache.maven.plugin.MojoExecutionException;
 import org.apache.maven.plugin.MojoFailureException;
 import org.apache.maven.plugins.annotations.Parameter;
@@ -70,6 +72,12 @@ public abstract class UploadSchemaRegistryMojo extends SchemaRegistryMojo {
     errors = 0;
     failures = 0;
 
+    if (decodeSubject) {
+      subjects = decode(subjects);
+      schemaTypes = decode(schemaTypes);
+      references = decode(references);
+    }
+
     for (String subject : subjects.keySet()) {
       processSubject(subject, false);
     }
@@ -107,11 +115,7 @@ public abstract class UploadSchemaRegistryMojo extends SchemaRegistryMojo {
         return;
       }
 
-      String subject = key;
-      if (decodeSubject && subject.contains(PERCENT_REPLACEMENT)) {
-        subject = decode(subject);
-      }
-      boolean success = processSchema(subject, file, schema.get(), schemaVersions);
+      boolean success = processSchema(key, file, schema.get(), schemaVersions);
       if (!success) {
         failures++;
       }
@@ -123,10 +127,21 @@ public abstract class UploadSchemaRegistryMojo extends SchemaRegistryMojo {
     }
   }
 
-  protected static String decode(String subject) throws UnsupportedEncodingException {
-    // Replace _x colon with percent sign, since percent is not allowed in XML name
-    subject = subject.replaceAll(PERCENT_REPLACEMENT, "%");
-    return URLDecoder.decode(subject, "UTF-8");
+  protected static <V> Map<String, V> decode(Map<String, V> map) {
+    return map.entrySet().stream()
+        .collect(Collectors.toMap(
+            e -> e.getKey().contains(PERCENT_REPLACEMENT) ? decode(e.getKey()) : e.getKey(),
+            Entry::getValue));
+  }
+
+  protected static String decode(String subject) {
+    try {
+      // Replace _x colon with percent sign, since percent is not allowed in XML name
+      subject = subject.replaceAll(PERCENT_REPLACEMENT, "%");
+      return URLDecoder.decode(subject, "UTF-8");
+    } catch (UnsupportedEncodingException e) {
+      throw new RuntimeException(e);
+    }
   }
 
   protected abstract boolean processSchema(String subject,

--- a/maven-plugin/src/test/java/io/confluent/kafka/schemaregistry/maven/SchemasWithDependenciesTest.java
+++ b/maven-plugin/src/test/java/io/confluent/kafka/schemaregistry/maven/SchemasWithDependenciesTest.java
@@ -120,4 +120,41 @@ public class SchemasWithDependenciesTest extends SchemaRegistryTest {
         Assert.assertNotNull("The schema should've been generated", pizza);
         Assert.assertTrue("The schema should contain fields from the dependency", pizza.toString().contains("currency"));
     }
+
+    @Test
+    public void testSchemaWithDependenciesWithSlash() throws Exception {
+        RegisterSchemaRegistryMojo schemaRegistryMojo = new RegisterSchemaRegistryMojo();
+        schemaRegistryMojo.client = new MockSchemaRegistryClient();
+
+        File pizzaFile = new File(tempDirectory, "pizza.avsc");
+        File amountFile = new File(tempDirectory, "amount.avsc");
+        try (
+            FileWriter pizzaWriter = new FileWriter(pizzaFile);
+            FileWriter amountWriter = new FileWriter(amountFile)
+        ) {
+            pizzaWriter.write(schema);
+            amountWriter.write(dependency);
+        }
+
+        Map<String, List<Reference>> schemaRefs = new LinkedHashMap<>();
+        List<Reference> subjectRefs = new ArrayList<>();
+        Reference ref = new Reference();
+        ref.name = "com.pizza.Amount";
+        ref.subject = "Root/Amount";
+        subjectRefs.add(ref);
+        schemaRefs.put("Root_x2FPizza", subjectRefs);
+        schemaRegistryMojo.references = schemaRefs;
+
+        Map<String, File> schemas = new LinkedHashMap<>();
+        schemas.put("Root_x2FAmount", amountFile);
+        schemas.put("Root_x2FPizza", pizzaFile);
+        schemaRegistryMojo.subjects = schemas;
+
+        schemaRegistryMojo.execute();
+        Schema pizza = ((AvroSchema) schemaRegistryMojo.schemas.get("Root/Pizza")).rawSchema();
+
+        Assert.assertNotNull("The schema should've been generated", pizza);
+        Assert.assertTrue("The schema should contain fields from the dependency", pizza.toString().contains("currency"));
+    }
+
 }

